graphbolt/src/cuda/index_select_impl.cu

```diff
@@ -115,7 +146,8 @@ torch::Tensor UVAIndexSelectImpl_(torch::Tensor input, torch::Tensor index) {
 
   // Sort the index to improve the memory access pattern.
   torch::Tensor sorted_index, permutation;
-  std::tie(sorted_index, permutation) = torch::sort(index);
+  std::tie(sorted_index, permutation) =
+      Sort(index, cuda::NumberOfBits(input_len));
```

Comment:
Could you make index sorting as an optional optimization for `IndexSelect`? From my profiling, index sorting could slow down UVA index selections in random cases.

Reply:
Instead of measuring throughput, can you measure the runtime? For the small sizes such as 1, 10, or even 1000 and 10000, the runtime is mostly identical and we are measuring the kernel launch latency, which is <50 microseconds according to my measurements. The benchmark should be performed with much larger-sized indices (>100000). If sort is disabled, then performance will be up to 4x worse than not sorting in some cases as seen below:

* params: n_rows=100000000, feat_size=47, num_indices=1000000, feature_device=Device.Pinned, indices_device=Device.GPU
|            |USE_ALIGN=1 |USE_ALIGN=0 |
|------------|------------|------------|
|USE_PERM=1  |   18966.435|   19071.299|
|USE_PERM=0  |    4389.194|    4390.788|

Paying around 40 microseconds for consistent overall performance is a big win in my opinion.

Reply:
If you report the runtime instead of the throughput in your profiling, you should see what I mean.

Reply:
I have profiled the whole DGL code before using the NVIDIA Nsight systems profiling tool. The Python overhead is a much bigger enemy when it comes to CUDA programming than a dozen microseconds of sort overhead.

Reply:
I am not quite sure about what you mean by "profiling runtime". I tried this setting on T4 but cannot reproduce your results. Here is what I got.

* params:  n_rows=100000000, feat_size=47, num_indices=1000000, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_ALIGN=1 |USE_ALIGN=0 |
|------------|------------|------------|
|USE_PERM=1  |    4665.199|    4671.809|
|USE_PERM=0  |    4751.018|    4750.682|

I think it is because the used GPU is different to your setting.

> Paying around 40 microseconds for consistent overall performance and up to 4x performance improvement is a big win in my opinion.

When I said index sorting could be slow, I didn't refer to the latency. I mean in some of my previous test cases, the throughput of index selection is negatively affected by the index sorting, especially for those random cases. So it is better for me to make index sorting as an optional optimization, even you may turn it on by default.

Reply:
Could you give me a few examples where sorting is significantly slower, where num_indices >= 100000?

Reply:
What I am trying to say is that in your tests, the work to be performed is so little, they mostly measured latency. Since sorting adds some latency, sorting will look slow. If you perform tests with enough work, I don't expect sorting to be slower in no circumstances.

Reply:
Maybe we can test with at least 8MB of data copied to ensure there is always enough work?

Reply:
By profiling runtime, I had meant reporting runtime numbers instead of throughput numbers.

Reply:
By the way, what is your GPU's PCIe setup? PCIe 3.0 16x should be around 12GB/s. I feel your setup's bandwidth is limited so the sorting benefits don't surface for you.

Reply:
Could you take a look at the updated description of this PR and #6644. As I said, for `num_indices=1` or `1000`, the runtime is less than 100us.

Reply:
This is what I got with indices = 100000.

* params:  n_rows=10000000, feat_size=64, num_indices=100000, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_ALIGN=1 |USE_ALIGN=0 |
|------------|------------|------------|
|USE_PERM=1  |    5685.875|    5719.948|
|USE_PERM=0  |    5918.350|    5909.524|

Index sorting is a bit slower in this case.

* params:  n_rows=100000, feat_size=64, num_indices=100000, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_ALIGN=1 |USE_ALIGN=0 |
|------------|------------|------------|
|USE_PERM=1  |    8593.935|    8821.381|
|USE_PERM=0  |    5926.886|    5927.017|

It has a better performance on a smaller size of feature table. 

> By the way, what is your GPU's PCIe setup?

My machine has a single T4 GPU on PCIe 3.0. 

So my feeling is the performance of index sorting depends on the datasets and hardware settings. And I believe it is beneficial for most of the skewed graphs. But since it could be slow on some cases, e.g., purely random access or small workloads, it is better to make this optimization configurable.